src/StoreClient.h

```diff
@@ -67,17 +68,23 @@ class store_client
 public:
     store_client(StoreEntry *);
     ~store_client();
-    bool memReaderHasLowerOffset(int64_t) const;
+    /// \returns the smallest of the passed value and the memory offset of the client
+    int memHeaderOffsetLowerThan(const int64_t offset) const;
     int getType() const;
     void fail();
     void callback(ssize_t len, bool error = false);
+    void callbackClientSide();
     void doCopy (StoreEntry *e);
     void readHeader(const char *buf, ssize_t len);
     void readBody(const char *buf, ssize_t len);
     void copy(StoreEntry *, StoreIOBuffer, STCB *, void *);
     void dumpStats(MemBuf * output, int clientNumber) const;
+    /// whether doCopy() can be called
+    bool canCopy() const { return canScheduleCallback() && !flags.disk_io_pending; }
```

Comment:
doCopy() is not documented so it is difficult for a reader to understand what "doCopy() can be called" really means. I think we know enough in this area to describe this new method in a more _direct_ way. Please try.

Reply:
We can define this method as 'returning true when a previous copy from store to client operation has been finished already (or no copy operation has been performed yet)' or 'returning false if a copy from store to client operation is being in progress right now'.

Reply:
> false if a copy ... operation is being in progress right now

This is a very good start, thank you! Is there a small set of (preferably public) methods that start a copy operation? That (upon return) finish a started copy operation? That abort a started copy operation? Depending on your answers, we may be able to clarify what that "copy operation" means.

Reply:
> clarify what that "copy operation" means

Basing on the canCopy() implementation, the 'copy operation' starts when either the swapin starts (if any) or CallbackClientSide() call is scheduled (in any case) - making clientSideCaller non-nil. In most cases (except one, IIRC storeSwapInFileClosed()), this happens in storeClientCopy2() or its descendants. It finishes (IMO) when the data were written to the client and (by a callback from client-side), another STCB passed to store_client::copy(). Within this period of time, another canCopy() will return false.